detector/calorimeter/dual-readout/src/DRconstructor.cpp

```diff
@@ -0,0 +1,440 @@
+
+bool ddDRcalo::DRconstructor::checkContained(TGeoTrap* rootTrap, dd4hep::Position& pos, double z, bool throwExcept) {
+  double pos_[3] = {pos.x(),pos.y(),z};
+  bool check = rootTrap->Contains(pos_);
+
+  if ( throwExcept && !check ) throw std::runtime_error("Fiber must be in the tower!");
+  return check;
+}
+
+void ddDRcalo::DRconstructor::getNormals(TGeoTrap* rootTrap, int numxBl2, double z, double* norm1, double* norm2, double* norm3, double* norm4) {
+  dd4hep::Position pos1 = dd4hep::Position( fSegmentation->localPosition(fNumx,fNumy,fNumx/2,0) );
+  dd4hep::Position pos2 = dd4hep::Position( fSegmentation->localPosition(fNumx,fNumy,fNumx/2+numxBl2/2-1,fNumy/2) );
+  dd4hep::Position pos3 = dd4hep::Position( fSegmentation->localPosition(fNumx,fNumy,fNumx/2,fNumy-1) );
+  dd4hep::Position pos4 = dd4hep::Position( fSegmentation->localPosition(fNumx,fNumy,fNumx/2-numxBl2/2+1,fNumy/2) );
+  double pos1_[3] = {pos1.x(),pos1.y(),z};
+  double pos2_[3] = {pos2.x(),pos2.y(),z};
+  double pos3_[3] = {pos3.x(),pos3.y(),z};
+  double pos4_[3] = {pos4.x(),pos4.y(),z};
+  double dir[3] = {0.,0.,0.};
+
+  rootTrap->ComputeNormal(pos1_,dir,norm1);
+  rootTrap->ComputeNormal(pos2_,dir,norm2);
+  rootTrap->ComputeNormal(pos3_,dir,norm3);
+  rootTrap->ComputeNormal(pos4_,dir,norm4);
+  norm1[2] = 0.; // check horizontal distance only
+  norm2[2] = 0.;
+  norm3[2] = 0.;
+  norm4[2] = 0.;
+}
+
+dd4hep::Box ddDRcalo::DRconstructor::calculateFullBox(TGeoTrap* rootTrap, int& rmin, int& rmax, int& cmin, int& cmax, double dz) {
+  float gridSize = fX_dim.distance();
+  double zmin = -rootTrap->GetDz() + TGeoShape::Tolerance();
+  float xmin = 0., ymin = 0., ymax = 0.;
+
+  for (int row = 0; row < fNumy; row++) { // bottom-up
+    auto localPosition = dd4hep::Position( fSegmentation->localPosition(fNumx,fNumy,fNumx/2,row) );
+    auto pos = localPosition + dd4hep::Position(0.,-gridSize/2.,0.);
+    if ( checkContained(rootTrap,pos,zmin) ) {
+      ymin = pos.y();
+      rmin = row;
+      break;
+    }
+  }
+
+  for (int row = fNumy-1; row !=0 ; row--) { // top-down
+    auto localPosition = dd4hep::Position( fSegmentation->localPosition(fNumx,fNumy,fNumx/2,row) );
+    auto pos = localPosition + dd4hep::Position(0.,gridSize/2.,0.);
+    if ( checkContained(rootTrap,pos,zmin) ) {
+      ymax = pos.y();
+      rmax = row;
+      break;
+    }
+  }
+
+  for (int col = 0; col < fNumx; col++) { // left-right
+    auto localPosition = dd4hep::Position( fSegmentation->localPosition(fNumx,fNumy,col,rmin) );
+    auto pos = localPosition + dd4hep::Position(-gridSize/2.,-gridSize/2.,0.);
+    if ( checkContained(rootTrap,pos,zmin) ) {
+      xmin = pos.x();
+      cmin = col;
+      break;
+    }
+  }
+
+  // assume phi symmetry
+  float xmax = -xmin;
+  cmax = fNumx-1 - cmin;
+
+  // verify assumptions
+  if ( std::abs(ymax+ymin) > TGeoShape::Tolerance() )
+    throw std::runtime_error("Envelop of full length fibers (fullBox) is not located at the centre of the tower!");
```

Comment:
seems like [this commit](https://github.com/HEP-FCC/dual-readout/commit/fb830a0390184fe99b79d094d5849bcce2e47caa) still hasn't been pulled?